README.md

<img src="https://github.com/tiagosito/grouping/raw/master/doc/assets/grouping_logo_full.png" width="100%" alt="logo" />
<h2 align="center">
  A component for you to group similar content, being able to add new, select or remove existing ones.
</h2>

<p align="center">
<a href="https://pub.dartlang.org/packages/grouping">
  <img alt="Pub Package" src="https://img.shields.io/pub/v/grouping.svg">
</a>
  <a href="https://www.buymeacoffee.com/tiagosito" target="_blank"><img alt="Buy Me A Coffee" src="https://i.imgur.com/aV6DDA7.png" </a>
</p>
 
  
## Overview

A simple data grouper

#### Contributing
  - [https://github.com/tiagosito/grouping](https://github.com/tiagosito/grouping)


#### Getting Started


In _pubspec.yaml_:
```yaml
dependencies:
  grouping: any
```

### Adding, Select and Removing data :
<img src="https://github.com/tiagosito/grouping/blob/master/doc/assets/grouping.gif?raw=true" width="30%" alt="Using grouping" />

```dart
import 'package:grouping/grouping.dart';

Grouping(
label: Text('Category'.toUpperCase(), style: TextStyle(color: Theme.of(context).primaryColor)),
iconAdd: Icon(
  Icons.add,
  size: 12,
  color: Theme.of(context).primaryColor,
),
labelAdd: Text('Add new category', style: TextStyle(color: Theme.of(context).primaryColor)),
height: 350,
//background: Colors.red,
items: _listItemsGroup,
onTap: (item) async {
  print('Click do Item ${item.toString()}');

  return item;
},
onRemove: <ModelExample>(item) async {
  //Simulating a call to the server
  var requestHTTP = await Future.delayed(Duration(seconds: 3));

  Future<bool> result = requestHTTP != null ? Future.value(true) : Future.value(false);

  print('Tentativa de Remover item ${item.toString()} = $result');
  return result;
},
onAdd: <ItemGroup>() async {
  ModelExample result = await Navigator.push(
    context,
    MaterialPageRoute(
      builder: (context) => TryGetValue(),
    ),
  );

  if (result == null) {
    print('No items were selected');
  } else {
    return Future.value(buildItemGroup(result));
  }
  return null;
},
),

.
.
.

buildItemGroup(ModelExample result) {
  return GroupingItem<ModelExample>(
    icon: Icon(Icons.payment, size: 35, color: Theme.of(context).primaryColor),
    remove: Icon(Icons.delete, size: 15, color: Theme.of(context).primaryColor),
    title: Text(
      '${result.name.toUpperCase()} - ${result.age}',
      style: TextStyle(color: Theme.of(context).primaryColor, fontWeight: FontWeight.bold),
    ),
    titleLeading: Text('due date: ${(DateTime.now().second % 2 == 0) ? '09/27/22' : '12/27/22'}', textAlign: TextAlign.right),
    subtitle: Text(
      (DateTime.now().second % 2 == 0)
          ? '256789.48592.70.9874 0 11'
          : '000000.00000.00.0000 0 00 00000 000000 000 0 00000 0 000',
      style: TextStyle(color: Colors.grey, fontWeight: FontWeight.normal),
    ),
    subtitleLeading: Text((DateTime.now().second % 2 == 0) ? 'U\$ 777.777.77' : 'U\$ 999.999.99', textAlign: TextAlign.right),
    item: result,
    //showRemoveButton: true, // default == true
    // background: Colors.grey[500],
  );
}
```

##### Model example
```dart
class ModelExample {
  final String name;
  final int age;

  ModelExample({this.name, this.age});

  @override
  String toString() {
    return '$name $age';
  }

  factory ModelExample.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;
    return ModelExample(
      name: json["name"],
      age: json["age"],
    );
  }

  static List<ModelExample> fromJsonList(List list) {
    if (list == null) return null;
    return list.map((item) => ModelExample.fromJson(item)).toList();
  }
}
```

## Features and bugs

Please send feature requests and bugs at the [issue tracker](https://github.com/tiagosito/grouping/issues).


CHANGELOG.md
# [0.0.2]

The label will be null by default

## [0.0.1] - Initial release

Initial release.


test/grouping_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:grouping/grouping.dart';

void main() {
  final grouping = Grouping(
    items: GroupItemList(items: []),
    onTap: (item) {
      return item;
    },
    onRemove: <ModelExample>(item) async {
      //Simulating a call to the server
      var requestHTTP = await Future.delayed(Duration(seconds: 3));

      Future<bool> result = requestHTTP != null ? Future.value(true) : Future.value(false);

      print('Tentativa de Remover item ${item.toString()} = $result');
      return result;
    },
    onAdd: <ItemGroup>() async {
      var result = await Future.delayed(Duration(milliseconds: 350));

      if (result == null) {
        print('No items were selected');
      } else {
        return Future.value(buildGroupItem());
      }
      return null;
    },
  );

  test('Items is not null', () {
    expect(grouping.items != null, true);
    expect(grouping.items.getListItems.length, 0);
  });
  test('Items == 0', () {
    expect(grouping.items.getListItems.length, 0);
  });

  test('onTap is not null', () {
    expect(grouping.onTap != null, true);
  });

  test('onRemove is not null', () {
    expect(grouping.onRemove != null, true);
  });

  test('onAdd is not null', () {
    expect(grouping.onAdd != null, true);
  });
}

buildGroupItem() {
  return GroupingItem<ModelExample>(
    title: Text(
      'Name',
    ),
    item: ModelExample(name: 'Tiago', age: 36),
  );
}

class ModelExample {
  final String name;
  final int age;

  ModelExample({this.name, this.age});

  @override
  String toString() {
    return '$name $age';
  }

  factory ModelExample.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;
    return ModelExample(
      name: json["name"],
      age: json["age"],
    );
  }

  static List<ModelExample> fromJsonList(List list) {
    if (list == null) return null;
    return list.map((item) => ModelExample.fromJson(item)).toList();
  }

  @override
  operator ==(object) => this.name.toLowerCase().contains(object.toLowerCase()) || this.age.toString().contains(object);

  @override
  int get hashCode => name.hashCode ^ age.hashCode;
}


lib/core/grouping.dart
import 'package:flutter/material.dart';
import 'package:grouping/model/grouping_item_list.dart';
import 'package:grouping/model/grouping_item.dart';

class Grouping<T> extends StatefulWidget {
  const Grouping({
    Key key,
    double height,
    Text label,
    Color background,
    Icon iconAdd,
    @required this.onAdd,
    Text labelAdd,
    @required this.onTap,
    @required this.onRemove,
    @required this.items,
  })  : this.height = height != null ? height : 150,
        this.label = label != null ? label : null,
        this.background =
            background != null ? background : const Color(0xFFF5F5F5),
        this.labelAdd = labelAdd != null ? labelAdd : const Text('CATEGORY'),
        assert(onAdd != null, 'onAdd Function cannot be null'),
        this.iconAdd = iconAdd != null
            ? iconAdd
            : const Icon(Icons.add, size: 12, color: Colors.red),
        assert(onRemove != null, 'onRemove Function cannot be null'),
        super(key: key);

  ///Property name [height]:
  ///
  ///Value type: double
  ///
  ///Maximum component height
  ///
  ///Default value is 150.0
  ///
  final double height;

  ///Property name label:
  ///
  ///Tag that best describes your data
  ///
  ///Value type: Text
  ///
  ///Default value: const Text('CATEGORY'),
  ///
  final Text label;

  ///Property name [background]:
  ///
  ///Component background color
  ///
  ///Value type: Color
  ///
  ///Default value: const Color(0xFFF5F5F5),
  ///
  final Color background;

  ///Property name [iconAdd]:
  ///
  ///add item icon
  ///
  ///Value type: Icon
  ///
  ///Default value: const Icon(Icons.add, size: 12, color: Colors.red),
  ///
  final Icon iconAdd;

  ///Property name [onAdd]:
  ///
  ///Add item Function
  ///
  ///This property is required, that is, it cannot be null
  ///
  ///Value type: Function<T>()
  ///
  ///Expected return: Future<GroupingItem<T>>
  ///
  ///Follow the example:
  ///
  ///```dart
  ///onAdd: <ItemGroup>() async {
  ///  ModelExample result = await Navigator.push(
  ///    context,
  ///    MaterialPageRoute(
  ///      builder: (context) => TryGetValue(),
  ///    ),
  ///  );
  ///
  ///  if (result == null) {
  ///    print('No items were selected');
  ///  } else {
  ///    return Future.value(buildItemGroup(result));//Return function
  ///  }
  ///  return null;
  ///},
  ///
  ///```
  ///
  ///```dart
  ///
  ///buildItemGroup(ModelExample result) {//Return function
  ///  return GroupingItem<ModelExample>(
  ///    icon: Icon(Icons.payment, size: 35, color: Theme.of(context).primaryColor),
  ///    remove: Icon(Icons.delete, size: 15, color: Theme.of(context).primaryColor),
  ///    title: Text(
  ///      '${result.name.toUpperCase()} - ${result.age}',
  ///      style: TextStyle(color: Theme.of(context).primaryColor, fontWeight: FontWeight.bold),
  ///    ),
  ///    titleLeading: Text('due date: 12/27/22', textAlign: TextAlign.right),
  ///    subtitle: Text(
  ///      '000000.00000.00.0000 0 00 00000 000000 000 0 00000 0 000',
  ///      style: TextStyle(color: Colors.grey, fontWeight: FontWeight.normal),
  ///    ),
  ///    subtitleLeading: Text('U\$ 999.999.99', textAlign: TextAlign.right),
  ///    item: result,
  ///    background: Colors.grey[50],
  ///  );
  ///}
  ///```
  ///
  final Future<GroupingItem<T>> Function<T>() onAdd;

  ///Property name [labelAdd]:
  ///
  ///Label describes the data that will be added
  ///
  ///Value type: Text
  ///
  ///Default value: null
  ///
  final Text labelAdd;

  ///Property name [onTap]:
  ///
  ///Tap item Function
  ///
  ///This property is required, that is, it cannot be null
  ///
  ///Value type: Function(T)
  ///
  ///Expected return: Future<T>
  ///
  ///Follow the example:
  ///
  ///```dart
  ///onTap: (item) async {
  ///  print('Click do Item ${item.toString()}');
  ///  return item;
  ///},
  ///```
  ///
  final Future<T> Function(T) onTap;

  ///Property name [onRemove]:
  ///
  ///Remove item Function
  ///
  ///This property is required, that is, it cannot be null
  ///
  ///Value type: Function<T>(T)
  ///
  ///Expected return: Future<bool>
  ///
  ///Follow the example:
  ///
  ///```dart
  ///onRemove: <ModelExample>(item) async {
  ///  //Simulating a call to the server
  ///  var requestHTTP = await Future.delayed(Duration(seconds: 3));
  ///
  ///  Future<bool> result = requestHTTP != null ? Future.value(true) : Future.value(false);
  ///
  ///  print('Tentativa de Remover item ${item.toString()} = $result');
  ///  return result;
  ///},
  ///```
  ///
  final Future<bool> Function<T>(T) onRemove;

  ///Property name [items]:
  ///
  ///A list of items to be added to the grouping
  ///
  ///Value type: [GroupItemList]
  ///
  final GroupItemList items;

  @override
  _GroupingState<T> createState() => _GroupingState<T>();
}

class _GroupingState<T> extends State<Grouping<T>> {
  ScrollController _scrollController = ScrollController();

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    _scrollController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Container(
        constraints: BoxConstraints(maxHeight: widget.height),
        color: widget.background,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (widget.label != null)
              Row(
                children: [
                  Expanded(child: widget.label),
                ],
              ),
            if (widget.label != null)
              SizedBox(
                height: 4.0,
              ),
            ValueListenableBuilder(
              valueListenable: widget.items,
              builder: (context, _, __) {
                return ConstrainedBox(
                  constraints: BoxConstraints(maxHeight: widget.height - 80),
                  child: ListView.builder(
                    shrinkWrap: true,
                    controller: _scrollController,
                    itemCount: widget.items.getListItems.length,
                    itemBuilder: (context, index) {
                      var item = widget.items.getListItems[index];
                      return Material(
                        color: item.background,
                        child: InkWell(
                          onTap: () async {
                            if (item.getIsRemoving()) {
                              debugPrint(
                                  '${item.item.toString()} is being removed');
                              return;
                            }
                            debugPrint(item.item.toString());
                            await widget.onTap(item.item);
                          },
                          child: Container(
                            child: Padding(
                              padding:
                                  const EdgeInsets.symmetric(vertical: 8.0),
                              child: Stack(
                                alignment: Alignment.centerLeft,
                                children: [
                                  if (item.getIsRemoving())
                                    Container(
                                      child: Center(
                                        child: CircularProgressIndicator(
                                          valueColor: AlwaysStoppedAnimation<
                                                  Color>(
                                              Theme.of(context).primaryColor),
                                        ),
                                      ),
                                    ),
                                  if (item.icon != null)
                                    ConstrainedBox(
                                      constraints: BoxConstraints(
                                          minWidth: 35, maxWidth: 35),
                                      child: item.icon,
                                    ),
                                  if (item.showRemoveButton)
                                    Positioned(
                                      top: 0,
                                      right: 0,
                                      child: Material(
                                        color: Colors.transparent,
                                        borderRadius:
                                            BorderRadius.circular(35.0),
                                        child: InkWell(
                                          highlightColor:
                                              item.removeHighlightColor,
                                          borderRadius:
                                              BorderRadius.circular(35.0),
                                          onTap: () async {
                                            if (item.getIsRemoving()) {
                                              return;
                                            }

                                            try {
                                              item.setIsRemoving(item);
                                              widget.items.updateList();
                                              var result = await widget
                                                  .onRemove(item.item);
                                              if (result == null) {
                                                return;
                                              } else {
                                                widget.items.removeItem(item);
                                              }
                                            } catch (e) {
                                              debugPrint(e);
                                            }
                                          },
                                          child: ConstrainedBox(
                                            constraints: BoxConstraints(
                                              minWidth: 35,
                                              maxWidth: 35,
                                              minHeight: 35,
                                              maxHeight: 35,
                                            ),
                                            child: Container(
                                              decoration: BoxDecoration(
                                                borderRadius:
                                                    BorderRadius.circular(35.0),
                                              ),
                                              child: Center(child: item.remove),
                                            ),
                                          ),
                                        ),
                                      ),
                                    ),
                                  Padding(
                                    padding: EdgeInsets.only(
                                        left: item.icon == null ? 0 : 50,
                                        right: item.remove == null ? 0 : 40),
                                    child: Column(
                                      children: [
                                        Row(
                                          mainAxisAlignment:
                                              MainAxisAlignment.spaceBetween,
                                          children: [
                                            Expanded(
                                              child: Padding(
                                                padding: EdgeInsets.symmetric(
                                                    vertical: (item
                                                                .showRemoveButton &&
                                                            (item.subtitle ==
                                                                    null ||
                                                                item.subtitle ==
                                                                    null))
                                                        ? 9.0
                                                        : 0.0),
                                                child: item.title,
                                              ),
                                            ),
                                            if (item.titleLeading != null)
                                              Padding(
                                                padding:
                                                    const EdgeInsets.symmetric(
                                                        horizontal: 8.0),
                                                child: ConstrainedBox(
                                                  constraints: BoxConstraints(
                                                      maxWidth: 120),
                                                  child: item.titleLeading,
                                                ),
                                              ),
                                          ],
                                        ),
                                        if (item.subtitle != null)
                                          SizedBox(
                                            height: 4,
                                          ),
                                        Row(
                                          mainAxisAlignment:
                                              MainAxisAlignment.spaceBetween,
                                          children: [
                                            if (item.subtitle != null)
                                              Expanded(
                                                child: item.subtitle,
                                              ),
                                            if (item.subtitleLeading != null)
                                              item.subtitle == null
                                                  ? Expanded(
                                                      child: Padding(
                                                        padding:
                                                            const EdgeInsets
                                                                    .symmetric(
                                                                horizontal:
                                                                    8.0),
                                                        child: ConstrainedBox(
                                                          constraints:
                                                              BoxConstraints(
                                                                  maxWidth:
                                                                      120),
                                                          child: item
                                                              .subtitleLeading,
                                                        ),
                                                      ),
                                                    )
                                                  : Padding(
                                                      padding: const EdgeInsets
                                                              .symmetric(
                                                          horizontal: 8.0),
                                                      child: ConstrainedBox(
                                                        constraints:
                                                            BoxConstraints(
                                                                maxWidth: 120),
                                                        child: item
                                                            .subtitleLeading,
                                                      ),
                                                    ),
                                          ],
                                        ),
                                      ],
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          ),
                        ),
                      );
                    },
                  ),
                );
              },
            ),
            //SizedBox(height: widget.items.getListItems == null || widget.items.getListItems.length == 0 ? 0.0 : 24.0),
            Material(
              color: widget.background,
              child: InkWell(
                onTap: () async {
                  var result = await widget.onAdd();

                  if (result != null) {
                    widget.items.setListItem(result);

                    if (widget.items.getListItems.length > 1) {
                      _scrollToEnd();
                    }
                  }
                },
                child: Container(
                  height: 35,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    mainAxisSize: MainAxisSize.max,
                    children: [
                      widget.iconAdd,
                      widget.labelAdd,
                    ],
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  _scrollToEnd() async {
    _scrollController.animateTo(
      _scrollController.position.maxScrollExtent + 100,
      duration: Duration(microseconds: 350),
      curve: Curves.fastOutSlowIn,
    );
  }
}


lib/model/grouping_item.dart
import 'package:flutter/material.dart';

class GroupingItem<T> {
  ///Property name [item]:
  ///
  ///Value type: T
  ///
  T item;

  ///Property name [icon]:
  ///
  ///Leading icon
  ///
  ///Value type: Icon
  ///
  Icon icon;

  ///Property name [remove]:
  ///
  ///Trailing icon
  ///
  ///Value type: Icon
  ///
  ///Default value: const Icon(Icons.delete, size: 12, color: Colors.red),
  ///
  Icon remove;

  ///Property name [removeHighlightColor]:
  ///
  /// The highlight color of the ink response when pressed. If this property is
  /// null then the highlight color of the theme, [ThemeData.highlightColor],
  /// will be used.
  ///
  /// Value type: Color
  ///
  /// Default value: [ThemeData.highlightColor],
  ///
  Color removeHighlightColor;

  ///Property name [title]:
  ///
  ///Tag that best describes your data
  ///
  ///This property is required, that is, it cannot be null
  ///
  Text title;

  ///Property name [titleLeading]:
  ///
  ///Title leading data
  ///
  Text titleLeading;

  ///Property name [subtitle]:
  ///
  ///Subtitle data
  ///
  Text subtitle;

  ///Property name [subtitle]:
  ///
  ///Subtitle leading data
  ///
  Text subtitleLeading;

  ///Property name [background]:
  ///
  ///Item background color
  ///
  ///Value type: Color
  ///
  Color background;

  ///Property name [hideRemoveButton]:
  ///
  ///Hide or show the Remove button
  ///
  ///Value type: bool
  ///
  /// Default value: true,
  ///
  bool showRemoveButton;

  ///Property name [_removing]:
  ///
  ///blocks functions when an item is being removed
  ///
  ///Internal use
  ///
  bool _removing;

  GroupingItem({
    T item,
    Icon icon,
    Icon remove,
    @required Text title,
    Text titleLeading,
    Text subtitle,
    Text subtitleLeading,
    Color background,
    bool showRemoveButton,
  }) {
    this.item = item;
    this.icon = icon;
    this.remove = remove != null
        ? remove
        : const Icon(Icons.delete, size: 12, color: Colors.red);
    this.removeHighlightColor = removeHighlightColor;
    this.title = title;
    this.titleLeading = titleLeading;
    this.subtitle = subtitle;
    this.subtitleLeading = subtitleLeading;
    this.background = background;
    this.showRemoveButton = showRemoveButton != null ? showRemoveButton : true;
    this._removing = false;
  }

  setIsRemoving(GroupingItem item) {
    item._removing = true;
  }

  bool getIsRemoving() => _removing;
}


lib/model/grouping_item_list.dart
import 'package:flutter/material.dart';

import 'grouping_item.dart';

class GroupItemList<T> extends ValueNotifier {
  GroupItemList({List<T> items}) : super(items) {
    if (items == null || items.length == 0) {
      return;
    } else {
      items.forEach((element) {
        setListItem((element as GroupingItem));
      });
    }
  }

  ///Property name [listItems] :
  ///
  ///A list to group [GroupingItem]
  ///
  ///Value type: ValueNotifier<List<GroupingItem>>
  ///
  /// Default value: [],
  ///
  ValueNotifier<List<GroupingItem>> listItems =
      ValueNotifier<List<GroupingItem>>([]);

  ///Metod name [getListItems] :
  ///
  ///A method that brings items from the ValueNotifier list [listItems]
  ///
  ///Note: can return null
  ///
  List<GroupingItem> get getListItems => listItems?.value ?? null;

  ///
  ///A method that inserts the data into the [listItems]
  ///
  ///Note: returns an exception if the [item] is null
  ///
  setListItem(GroupingItem item) {
    if (item == null) {
      return throw ('ItemGroup cannot be null or empty');
    }

    if (this.listItems == null) {
      List<GroupingItem> list = [item];
      this.listItems = ValueNotifier<List<GroupingItem>>(list);
    } else {
      var itemsTemp =
          this.listItems.value.where((element) => element == item).toList();
      if (itemsTemp != null && itemsTemp.length > 0) {
        this.listItems.value.remove(item);
      }
      this.listItems.value.add(item);
    }

    updateList();
  }

  ///
  ///A method that remove the data into the [listItems]
  ///
  ///Note: returns an exception if the [item] is null
  ///
  ///or if the [listItems] is null
  ///
  ///or if the [listItems].length is equal 0
  ///
  removeItem(T item) {
    if (item == null) {
      return throw ('Item cannot be null');
    }

    if (this.getListItems == null) {
      return throw ('The list cannot be null');
    }

    if (this.getListItems.length == 0) {
      return throw ('The list is already empty');
    }

    this.listItems.value.remove(item);

    updateList();
  }

  ///
  ///A method that update [listItems]
  ///
  updateList() {
    notifyListeners();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:grouping/grouping.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Grouping',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Group(),
    );
  }
}

class Group extends StatefulWidget {
  @override
  _GroupState createState() => _GroupState();
}

class _GroupState extends State<Group> {
  List<GroupingItem<ModelExample>> list = [];
  GroupItemList _listItemsGroup = GroupItemList(items: []);

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Grouping'),
      ),
      body: Column(
        children: [
          SizedBox(
            height: 8,
          ),
          Container(
            child: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Grouping(
                label: Text('Category'.toUpperCase(), style: TextStyle(color: Theme.of(context).primaryColor)),
                iconAdd: Icon(
                  Icons.add,
                  size: 12,
                  color: Theme.of(context).primaryColor,
                ),
                labelAdd: Text('Add new category', style: TextStyle(color: Theme.of(context).primaryColor)),
                height: 350,
                //background: Colors.red,
                items: _listItemsGroup,
                onTap: (item) async {
                  print('Click do Item ${item.toString()}');

                  return item;
                },
                onRemove: <ModelExample>(item) async {
                  //Simulating a call to the server
                  var requestHTTP = await Future.delayed(Duration(seconds: 1));

                  Future<bool> result = requestHTTP != null ? Future.value(true) : Future.value(false);

                  print('Tentativa de Remover item ${item.toString()} = $result');
                  return result;
                },
                onAdd: <ItemGroup>() async {
                  ModelExample result = await Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => TryGetValue(),
                    ),
                  );

                  if (result == null) {
                    print('No items were selected');
                  } else {
                    return Future.value(buildItemGroup(result));
                  }
                  return null;
                },
              ),
            ),
          ),
        ],
      ),
    );
  }

  buildItemGroup(ModelExample result) {
    return GroupingItem<ModelExample>(
      icon: Icon(Icons.payment, size: 35, color: Theme.of(context).primaryColor),
      remove: Icon(Icons.delete, size: 15, color: Theme.of(context).primaryColor),
      title: Text(
        '${result.name.toUpperCase()} - ${result.age}',
        style: TextStyle(color: Theme.of(context).primaryColor, fontWeight: FontWeight.bold),
      ),
      titleLeading: Text('due date: ${(DateTime.now().second % 2 == 0) ? '09/27/22' : '12/27/22'}', textAlign: TextAlign.right),
      subtitle: Text(
        (DateTime.now().second % 2 == 0)
            ? '256789.48592.70.9874 0 11'
            : '000000.00000.00.0000 0 00 00000 000000 000 0 00000 0 000',
        style: TextStyle(color: Colors.grey, fontWeight: FontWeight.normal),
      ),
      subtitleLeading: Text((DateTime.now().second % 2 == 0) ? 'U\$ 777.777.77' : 'U\$ 999.999.99', textAlign: TextAlign.right),
      item: result,
      //showRemoveButton: true, // default == true
      // background: Colors.grey[500],
    );
  }
}

class TryGetValue extends StatefulWidget {
  @override
  _TryGetValueState createState() => _TryGetValueState();
}

class _TryGetValueState extends State<TryGetValue> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Column(
        children: [
          RaisedButton(
            child: Text('Ok'),
            onPressed: () {
              var item = ModelExample(name: 'Tiago', age: DateTime.now().second);
              Navigator.pop(context, item);
            },
          )
        ],
      ),
    );
  }
}

class ModelExample {
  final String name;
  final int age;

  ModelExample({this.name, this.age});

  @override
  String toString() {
    return '$name $age';
  }

  factory ModelExample.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;
    return ModelExample(
      name: json["name"],
      age: json["age"],
    );
  }

  static List<ModelExample> fromJsonList(List list) {
    if (list == null) return null;
    return list.map((item) => ModelExample.fromJson(item)).toList();
  }

  @override
  operator ==(object) => this.name.toLowerCase().contains(object.toLowerCase()) || this.age.toString().contains(object);

  @override
  int get hashCode => name.hashCode ^ age.hashCode;
}
